doc/Authoring_Extensions.md

```diff
@@ -0,0 +1,250 @@
+
+## Prerequisites
+
+You’ll need node in version 8:
+
+```bash
+curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.5/install.sh | bash
+nvm install 8
+```
+
+and yarn in version 1.0.2
+
+```
+npm install -g yarn@1.0.2
+```
+
+Also make sure your `python --version` points to a Python 2.x installation.
+
+## Project Layout
+
+We are going to create a monorepo (a repository containing multiple npm packages) named `theia-hello-world-extension` containing three packages: `hello-world-extension`, `browser-app` and `electron-app`. The first contains our extension, the latter two the Theia applications to run our extension in browser and electron mode.  We are going to use `yarn` instead of `npm`, because it allows to structure such monorepos into workspaces. In our case, each workspace contains its own `npm` package. Common dependencies of these packages are 'hoisted' by `yarn` to their common root directory. We are also going to use `lerna` to run scripts accross workspaces. 
+
+The setup of such a repository is a bit tricky. This is why we have created a [code generator](https://www.npmjs.com/package/generator-theia-extension) to scaffold the project. It will also generate the `hello-world` example. Run it using
+
+```bash
+npm install -g yo generator-theia-extension
+mkdir theia-hello-world-extension
+cd theia-hello-world-extension
+yo theia-extension hello-world
+```
+
+Let's have look at the generated code now. The root `package.json` defines the workspaces, the dependency to `lerna` and some scripts to rebuild the native packages for browser or electron.
+
+```json
+{
+  "private": true,
+  "scripts": {
+    "prepare": "lerna run prepare",
+    "rebuild:browser": "theia rebuild:browser",
+    "rebuild:electron": "theia rebuild:electron"
+  },
+  "devDependencies": {
+    "lerna": "2.4.0"
+  },
+  "workspaces": [
+    "hello-world-extension", "browser-app", "electron-app"
+  ]
+}
+```
+
+We also got a `lerna.json` file to configure `lerna`:
+
+```json
+{
+  "lerna": "2.4.0",
+  "version": "0.1.0",
+  "useWorkspaces": true,
+  "npmClient": "yarn",
+  "command": {
+    "run": {
+      "stream": true
+    }
+  }
+}
+```
+
+## Implementing the Extension
+
+Next let's look at the generated code for our extension in the `hello-world-extension` folder. Let’s start with the `package.json`. It specifies the package’s metadata, its dependencies to the (bleeding edge) theia core package, a few scripts and dev dependencies, and the theia-extensions.
+
+The keyword `theia-extension` is important: It allows a Theia app to identify and install Theia extensions from `npm`.
```

Comment:
@akosyakov is that still the case? I thought we decided that existence of `theiaExtensions` is enough?

Reply:
it is important for the search, but is not required for install.

Reply:
so we cannot search for the existence of `theiaExtensions`?

Reply:
It would be quite expensive to request `npm view` for each package returned by the search. We could index all packages with npm registry web-hooks then the keyword will be redundant. Not sure how it will work for private registries.

Reply:
ok, understood